logstash-core/src/main/java/org/logstash/config/ir/compiler/EventCondition.java

```diff
@@ -0,0 +1,758 @@
+                if (lfound instanceof ConvertedList || lfound instanceof ConvertedMap) {
+                    return false;
+                } else if (lfound instanceof RubyString && rfound instanceof RubyString) {
+                    return ((RubyString) lfound).getByteList()
+                        .indexOf(((RubyString) rfound).getByteList()) > -1;
+                } else if (rfound instanceof ConvertedList) {
+                    return contains((ConvertedList) rfound, lfound);
+                } else {
+                    return lfound != null && rfound != null && lfound.equals(rfound);
+                }
+            }
+        }
+
+        private static final class FieldInConstantList implements EventCondition {
+
+            private final FieldReference field;
+
+            private final List<?> value;
+
+            private FieldInConstantList(final FieldReference field, final List<?> value) {
+                this.field = field;
+                this.value = value;
+            }
+
+            @Override
+            public boolean fulfilled(final JrubyEventExtLibrary.RubyEvent event) {
+                final Object found = event.getEvent().getUnconvertedField(field);
+                return found != null &&
+                    value.stream().anyMatch(val -> val.toString().equals(found.toString()));
+            }
+        }
+
+        private static final class FieldTruthy implements EventCondition {
+
+            private final FieldReference field;
+
+            private FieldTruthy(final FieldReference field) {
+                this.field = field;
+            }
+
+            @Override
+            public boolean fulfilled(final JrubyEventExtLibrary.RubyEvent event) {
+                final Object object = event.getEvent().getUnconvertedField(field);
+                if (object == null) {
+                    return false;
+                }
+                final String other = object.toString();
+                return other != null && !other.isEmpty() &&
+                    !Boolean.toString(false).equals(other);
+            }
+        }
+
+        /**
+         * This {@link IllegalArgumentException} is thrown when the inputs to an {@code if}
+         * condition do not conform to the expected types. It being thrown is a bug in Logstash
+         * in every case.
+         */
+        private static final class UnexpectedTypeException extends IllegalArgumentException {
+
+            private static final long serialVersionUID = 1L;
+
+            UnexpectedTypeException(final Expression left, final Expression right) {
+                super(
+                    String.format("Unexpected input types %s %s", left.getClass(), right.getClass())
+                );
+            }
+
+            UnexpectedTypeException(final Object inner) {
+                super(String.format("Unexpected input type %s", inner.getClass()));
+            }
+        }
+    }
```

Comment:
This is just at the compiler level. I think you can read from the stack-trace quite easily also this should be impossible to run into. If this is thrown there is a case missing in the compiler logic, a broken config can't cause this (if it can then I missed a something, that's why this is there :)).